ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs

```diff
@@ -179,63 +252,68 @@ data ChainSyncInstruction header point tip
+    => Tracer m TraceChainsyncClientObservation
+    -> ChainSyncClientPipelined header point tip m a
     -> PeerPipelined (ChainSync header point tip) AsClient StIdle m a
 
-chainSyncClientPeerPipelined (ChainSyncClientPipelined mclient) =
-    PeerPipelined $ SenderEffect $ chainSyncClientPeerSender Zero <$> mclient
+chainSyncClientPeerPipelined tracer (ChainSyncClientPipelined mclient) =
+    PeerPipelined $ SenderEffect $ chainSyncClientPeerSender tracer Zero <$> mclient
 
 
 chainSyncClientPeerSender
     :: forall n header point tip m a.
        Monad m
-    => Nat n
+    => Tracer m TraceChainsyncClientObservation
+    -> Nat n
     -> ClientPipelinedStIdle n header point tip m a
     -> PeerSender (ChainSync header point tip)
                   AsClient StIdle n
                   (ChainSyncInstruction header point tip)
                   m a
 
-chainSyncClientPeerSender n@Zero (SendMsgRequestNext stNext stAwait) =
-
-    SenderYield
+chainSyncClientPeerSender tracer n@Zero (SendMsgRequestNext stNext stAwait) = SenderEffect $ do
+    traceWith tracer $ NextObservation ObserveBefore
+    return $ SenderYield
       (ClientAgency TokIdle)
       MsgRequestNext
       (SenderAwait
         (ServerAgency (TokNext TokCanAwait))
           $ \case
             MsgRollForward header tip ->
-              SenderEffect $
-                  chainSyncClientPeerSender n
+              SenderEffect $ do
+                  traceWith tracer $ NextObservation ObserveAfter
+                  chainSyncClientPeerSender tracer n
                     <$> recvMsgRollForward header tip
                 where
                   ClientStNext {recvMsgRollForward} = stNext
 
             MsgRollBackward pRollback tip ->
-              SenderEffect $
-                  chainSyncClientPeerSender n
+              SenderEffect $ do
+                  traceWith tracer $ NextObservation ObserveAfter
+                  chainSyncClientPeerSender tracer n
                     <$> recvMsgRollBackward pRollback tip
                 where
                   ClientStNext {recvMsgRollBackward} = stNext
 
-            MsgAwaitReply ->
-              SenderAwait
+            MsgAwaitReply -> SenderEffect $ do
+              traceWith tracer $ NextObservation ObserveAfter
+              return $ SenderAwait
                 (ServerAgency (TokNext TokMustReply))
                 $ \case
                   MsgRollForward header tip ->
                     SenderEffect $ do
                       ClientStNext {recvMsgRollForward} <- stAwait
-                      chainSyncClientPeerSender n
+                      chainSyncClientPeerSender tracer n
                         <$> recvMsgRollForward header tip
 
                   MsgRollBackward pRollback tip ->
                     SenderEffect $ do
                       ClientStNext {recvMsgRollBackward} <- stAwait
-                      chainSyncClientPeerSender n
+                      chainSyncClientPeerSender tracer n
```

Comment:
To avoid passing tracer around we could create a closure which has access to it.

Reply:
I'm fine with passing the tracer around.